Add plugin: Macros

JamesCliffordSpratt opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/JamesCliffordSpratt/macros

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [x]  macOS
  - [x]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

Zachatoo commented:
["main": "main.js",](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/manifest.json#L7)
Do not use properties that are not documented in your plugin's [manifest](https://docs.obsidian.md/Reference/Manifest).

Why are you including the obsidian.d.ts file in your repo? You should already have access to those types through the obsidian package in your package.json file.

[this.addRibbonIcon('apple', 'Search for Food (FatSecret)', () => {](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/main.ts#L159), [name: 'Search for Food (FatSecret)',](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/main.ts#L182), [combinedHeaderCell.innerText = 'Combined Totals';](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/processors.ts#L290), [aggregateRow.insertCell().innerText = 'Aggregate Totals';](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/processors.ts#L536), [contentEl.createEl('h2', { text: 'Enter Food Search Term', cls: 'macros-modal-title' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L172), [text: \`Results for "${this.searchTerm}" (Page ${this.currentPage + 1})\`,](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L249), [text: 'New Meal Template',](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L306), [nameInput.placeholder = 'Meal Name (e.g., Meal1)';](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L315), [text: 'Create Meal Template',](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L343), [text: \`Edit Meal: ${this.meal.name}\`,](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L374), [text: '+ Add Food Item',](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L399), [text: \`Custom Serving Size for ${this.foodName}\`,](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L440), [text: \`Add Food Items to "${this.meal.name}"\`,](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L510), [dropdown.createEl('option', { text: '-- Select Food --', value: '' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L519), [const addBtn = row.createEl('button', { text: '+ Add Selected Item' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L525), [contentEl.createEl('h2', { text: 'Add to Macros', cls: 'macros-modal-title' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L48), [mealRow.createEl('label', { text: 'Select Meal:' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L52), [const addMealBtn = mealRow.createEl('button', { text: 'Add Meal' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L58), [foodRow.createEl('label', { text: 'Select Food:' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L72), [const addFoodBtn = foodRow.createEl('button', { text: 'Add Food' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L81), [text: 'Confirm Changes',](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/modals.ts#L135), [containerEl.createEl('h2', { text: 'Macros Plugin Settings' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L57), [.setName('Storage Folder')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L61), [.setName('Protein Color')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L82), [.setDesc('Color code for Protein slice (e.g., #4caf50)')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L83), [.setName('Fat Color')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L110), [.setDesc('Color code for Fat slice (e.g., #f44336)')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L111), [.setName('Carbs Color')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L138), [.setDesc('Color code for Carbs slice (e.g., #2196f3)')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L139), [previewContainer.createEl('h3', { text: 'Pie Chart Preview' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L166), [containerEl.createEl('h3', { text: 'Meal Templates' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L188), [.setName('Create a new Meal Template')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L190), [btn.setButtonText('+ Add Meal Template').onClick(() => {](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L193), [.setName('FatSecret API Key')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L245), [.setDesc('Enter your FatSecret API Key. Leave blank to use the default provided by the Macros Plugin.')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L246), [.setName('FatSecret API Secret')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L261), [.setDesc('Enter your FatSecret API Secret. Leave blank to use the default provided by the Macros Plugin.')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L262), [.setName('Test API Connection')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L275), [.setButtonText('Test Connection')](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L279)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[const content = await plugin.app.vault.read(activeFile);](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/processors.ts#L59), [const content = await this.app.vault.read(file);](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/main.ts#L476)
[Prefer to use `Vault.cachedRead()` over `Vault.read()` when you only want to read a file and not modify it](https://docs.obsidian.md/Plugins/Vault#Read+files)

[const exactMatches = files.filter(f => f.name.replace(/\.md$/, '').toLowerCase() === queryLower);](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/utils.ts#L191)
[Avoid iterating all files to find a file by its path](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+iterating+all+files+to+find+a+file+by+its+path)

[await plugin.app.vault.modify(activeFile, content);](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/processors.ts#L333)
[Prefer to use `Vault.process()` instead of `Vault.modify()` to modify a file in the background](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+%60Vault.process%60+instead+of+%60Vault.modify%60+to+modify+a+file+in+the+background)

[containerEl.createEl('h2', { text: 'Macros Plugin Settings' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L57)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

[containerEl.createEl('h2', { text: 'Macros Plugin Settings' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L57), [previewContainer.createEl('h3', { text: 'Pie Chart Preview' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L166), [containerEl.createEl('h3', { text: 'Meal Templates' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L188), [containerEl.createEl('h3', { text: 'Advanced' });](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L227)
For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```

[this.plugin.settings.storageFolder = value;](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L68)
[Use `normalizePath` to clean up user defined paths](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+%60normalizePath()%60+to+clean+up+user-defined+paths)

[new Setting(containerEl)](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L60-L78)
Prefer to use [AbstractInputSuggest](https://docs.obsidian.md/Reference/TypeScript+API/AbstractInputSuggest) when letting a user select a file/folder to add type-ahead support.

[new Setting(containerEl)](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L81-L106), [new Setting(containerEl)](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L109-L134), [new Setting(containerEl)](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/settings.ts#L137-L162)
Prefer to use [addColorPicker](https://docs.obsidian.md/Reference/TypeScript+API/Setting/addColorPicker) when letting a user select a color.

[const colorStyle = el.ownerDocument.createElement('style');](https://github.com/JamesCliffordSpratt/macros/blob/c380601a9b61081b288b849c86cceee9a3354a08/src/utils.ts#L333-L335)
This will leave behind style elements when your plugin is unloaded. Please ensure these are cleaned up.

JamesCliffordSpratt commented:
Thank you for your review. I have made the following changes:
- Version 2: Style overhaul and updated modular architecture
- Manifest File: Removed undocumented properties from manifest.json.
-  Type Definitions: Removed obsidian.d.ts from the repository.
- UI Text Formatting: Standardized all UI text to sentence case.
- File Reading Methods: Replaced Vault.read() with Vault.cachedRead() where appropriate.
- File Iteration: Optimized file lookup to avoid iterating over all files.
- File Modification: Switched from Vault.modify() to Vault.process() for background file modifications.
- Settings Tab Headings: Removed top-level headings from the settings tab.
- Settings Section Headings: Utilized new Setting(containerEl).setName('Section Name').setHeading(); for section headings.
- Path Normalization: Implemented normalizePath() for user-defined paths.
- Input Suggestions: Integrated AbstractInputSuggest for type-ahead support in file/folder selection.
- Color Picker Usage: Replaced manual color inputs with addColorPicker() for color selection.
- Style Element Cleanup: Ensured dynamic style elements are removed upon plugin unload.

I look forward to you additional review.

joethei commented:
[export class EventManager {](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/utils/EventManager.ts#L9-L56)
The Obsidian API already includes a `registerDomEvent` function for this purpose.

['Enter Food Search Term'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/FoodSearchModal.ts#L27), [`Edit Meal Template:](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/EditMealTemplateModal.ts#L24), [Select Food](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/EditMealTemplateModal.ts#L35), [Add Item](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/EditMealTemplateModal.ts#L42), [Custom Serving Size](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/CustomServingSizeModal.ts#L39), ['Add to Macros'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddToMacrosModal.ts#L26), ['Select Meal:'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddToMacrosModal.ts#L29), ['Add Meal'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddToMacrosModal.ts#L38),['Select Food:'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddToMacrosModal.ts#L53),['Add Food'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddToMacrosModal.ts#L68), [text: 'Confirm Changes',](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddToMacrosModal.ts#L135), ['New Meal Template'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddMealTemplateModal.ts#L26), ['Create Meal Template'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddMealTemplateModal.ts#L65), [Add Food Items](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddFoodToMealModal.ts#L29), [Select Food](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddFoodToMealModal.ts#L40), [Add Selected Item](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/modals/AddFoodToMealModal.ts#L47), ['Search for Food'](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/ui/live-search/LiveSearchModal.ts#L36)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[new Setting(containerEl).setName('🔧 Developer settings').setHeading();](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/settings/StorageService.ts#L377)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

[private DEFAULT_FAT_SECRET_API_KEY = '4430c3c828c04907bc8b559418435673';](https://github.com/JamesCliffordSpratt/macros/blob/3d1d0602779552ab409c9a3e66b5cdcceec60fcf/src/core/APIService.ts#L8-L9)
These should be configured by the user, you should never add API keys to source control.

JamesCliffordSpratt commented:
Thanks for your review. I have addressed all the issues you mentioned:

1. API Keys in Source Control:
- Removed all hardcoded API keys from APIService.ts
- API credentials are now required to be configured by users in settings
- Added clear UI indicators showing credential status (configured/not configured)
- Food search functionality handles missing credentials with helpful error messages
- Updated settings UI to explain the requirement for user-provided credentials

2. Obsidian's registerDomEvent:
- Removed custom EventManager class
- Updated all DOM event registrations to use Obsidian's built-in plugin.registerDomEvent() method

3. UI Text Case (Sentence Case):
- Updated all UI text to use proper sentence case

4. Settings Heading Text:
- Changed "🔧 Developer settings" to "🔧 Developer mode"
